### DISCUSSION TITLE
New release, wrt. to the Bipartite graph function

### DISCUSSION THREAD

**Discussion by Engberg-Jacob** [scipopt/PySCIPOpt#889] [General]
Hello! I can see that the Bipartite graph representation have been added to the next release of PySCIPOpt, which is something I could really use - does anyone know when the next release will be?

I've tried to build directly from that branch, but the SCIP opt Suite have been fighting me regarding being installed. I also tried looking into modifing PySCIPOpt after installation, but haven't had any luck/it were to complicated.

Maybe @Opt-Mucca would tell me the third-party software mentioned in #874?

(I've also looked at the package Ecole, but have met difficulties there, which I'll explained if needed)

Thank you for all your hard work!

**Comment by Opt-Mucca** [comment]
We're still not currently sure on the exact next release date. Sorry for the delay on this. We're currently stuck with some issues that would either mean releasing PySCIPOpt with a non-release-tag SCIP version, waiting for a new SCIP release, or just ignoring the issues for now. Would say we're experiencing a bit of decision paralysis.

For building SCIP: Do you need to actually build it from scratch? Can you download the pre-compiled versions from the main website: https://scipopt.org/index.php#download Those just need to be unpacked and you need some fairly basic system libraries. If you want to build against something else that's been pre-compiled then you can build against the SCIP libraries we use for releasing PySCIPOpt: https://github.com/scipopt/scipoptsuite-deploy/releases/tag/v0.4.0 From that link download `libscip-xxx` where `xxx` is your operating system. Extract the zip file and use it the location as `SCIPOPTDIR` when you're building PySCIPOpt. We'll be releasing some documentation soon that should help explain the build process a bit more.

The third-party software was simply Ecole, so I think you're already familiar with it. Very glad to hear that you find the new functionality helpful! That's incredibly rewarding to hear. I hope the above information helped (if not then feel free to shoot some more questions).

**Reply by Engberg-Jacob** [reply]
Yes! Perfect, thank you so much! (The SCIP library relases are SO much easier to work with, thank you).

I've got access to the function now. As I've read, the function needs to be called during the solving process, and needs a plugin to do so. Do you have a good resource to look at to create these plugins? I'm not familiar with extending big packages.
I'm not sure whether I'll need to do both during data generation, or just as a single call on an already defined model.

And thank you for the help again!

**Reply by Opt-Mucca** [reply]
So probably the best resource would be the test that got introduced alongside the feature. That adds the feature extractor to a branching rule (the link: https://github.com/scipopt/PySCIPOpt/blob/master/tests/test_bipartite.py ). It skips all the actual branching logic, but that can be seen from then looking some of the other branching tests, e.g. the strong branching test.

The most important piece of information to keep in mind is that the context of the current LP needs to be clear when you are using the bipartite graph representation. So in something like variable selection (branching) the current node and therefore the current LP is clear.

For extending this functionality to a research context: Data generation is entirely on your side. We can provide an easy function for a single instance and single LP (which we encourage you to use), but we can't have a single function for all possible representations.
For whether you need to use a single call or multiple: It depends on how many times you want to represent an instance. If you snapshot the instance just after presolving and the root node LP (before any cuts get added) then that could represent your instance. Does it continue to represent your instance fairly on the 500th node in the branch-and-bound tree? That's a currently not well understood question. Goodluck with work!

**Reply by Engberg-Jacob** [reply]
Alright, thank you! I think I'll be able to get what I want out of this (to extract a Bipartite Graph representation of a given MILP), thank you!

**Comment by Engberg-Jacob** [comment]
Sorry, I can see I never closed this, despite getting help, my bad! Thank you again!
